CI staging with rebase task can silently fail

## Description

For example, in https://circleci.com/gh/freedomofpress/securedrop/21323 , the full ci run does not occur. The gce-start script does not exit successfully, and the rest of the run (namely the tests) are not run, and it is not reported as a CI failure. For example:

```
make[1]: Entering directory `/home/circleci/sd'
./devops/gce-nested/gce-runner.sh
ssh: connect to host <snip> port 22: Connection timed out

rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]
make[1]: *** [ci-run] Error 255
```


## Expected Behavior
CI should fail when provisioning of CI instance fails

## Actual Behavior

CI passes
## Comments

Suggestions to fix, any other relevant information.

The failure is masked because the call to the failing function is wrapped in a bash if statement. To observe the behavior locally, apply this patch:

```
diff --git a/Makefile b/Makefile
index cd497579..d0ed592b 100644
--- a/Makefile
+++ b/Makefile
@@ -21,6 +21,22 @@ ci-run: ## Provisions GCE host for testing staging environment.
 ci-run-xenial: ## Provisions GCE host for testing staging environment.
        ./devops/gce-nested/gce-runner.sh xenial
 
+.PHONY: debug-ci-nested
+debug-ci-nested: ## Debug errors not propagating
+       @if [[ "${CIRCLE_BRANCH}" != docs-* ]]; then \
+               true; \
+               echo "true"; \
+               false; \
+               echo "false"; \
+       fi
+
+.PHONY: debug-ci-inline
+debug-ci-inline: ## Debug errors not propagating
+       true
+       echo "true"
+       false
+       echo "false"
+
 .PHONY: ci-go
 ci-go: ## Creates, provisions, tests, and destroys GCE host for testing staging environment.
        @if [[ "${CIRCLE_BRANCH}" != docs-* ]]; then \
```

Then run the targets in succession:

```
$ make debug-ci-nested
true
false
$ echo $?
0
$ make debug-ci-inline
true
echo "true"
true
false
Makefile:35: recipe for target 'debug-ci-inline' failed
make: *** [debug-ci-inline] Error 1
$ echo $?
2
```

Simply wrapping the call with a script containing `set -e` should resolve. We already have Circle CI re-running the destroy step explicitly to avoid long-running test instances.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -236,8 +236,8 @@ jobs:
           command: ./devops/scripts/rebase-ci.sh
 
       - run:
-          name: Run Staging tests on GCE
-          command: make ci-xenial-go
+          name: Run Staging tests on GCE (Xenial)
+          command: make ci-go-xenial
 
       - run:
           name: Ensure environment torn down

--- a/Makefile
+++ b/Makefile
@@ -5,37 +5,17 @@ PWD := $(shell pwd)
 TAG ?= $(shell git rev-parse HEAD)
 STABLE_VER := $(shell cat molecule/shared/stable.ver)
 
-.PHONY: ci-spinup
-ci-spinup: ## Creates GCE host for testing staging environment.
-	./devops/gce-nested/gce-start.sh
+.PHONY: ci-go
+ci-go: ## Creates, provisions, tests, and destroys GCE host for testing staging environment.
+	./devops/gce-nested/ci-go.sh
+
+.PHONY: ci-go-xenial
+ci-go-xenial: ## Creates, provisions, tests, and destroys GCE host for testing staging environment under xenial.
+	./devops/gce-nested/ci-go.sh xenial
 
 .PHONY: ci-teardown
 ci-teardown: ## Destroys GCE host for testing staging environment.
 	./devops/gce-nested/gce-stop.sh
-
-.PHONY: ci-run
-ci-run: ## Provisions GCE host for testing staging environment.
-	./devops/gce-nested/gce-runner.sh
-
-.PHONY: ci-run-xenial
-ci-run-xenial: ## Provisions GCE host for testing staging environment.
-	./devops/gce-nested/gce-runner.sh xenial
-
-.PHONY: ci-go
-ci-go: ## Creates, provisions, tests, and destroys GCE host for testing staging environment.
-	@if [[ "${CIRCLE_BRANCH}" != docs-* ]]; then \
-		make ci-spinup; \
-		make ci-run; \
-		make ci-teardown; \
-	fi
-
-.PHONY: ci-xenial-go
-ci-xenial-go: ## Creates, provisions, tests, and destroys GCE host for testing staging environment under xenial.
-	@if [[ "${CIRCLE_BRANCH}" != docs-* ]]; then \
-		make ci-spinup; \
-		make ci-run-xenial; \
-		make ci-teardown; \
-	fi
 
 .PHONY: ci-lint
 ci-lint: ## Runs linting in linting container.

--- a/devops/gce-nested/ci-go.sh
+++ b/devops/gce-nested/ci-go.sh
@@ -1,0 +1,26 @@
+#!/bin/bash
+# Wrapper script to managed GCE-based CI runs, specifically:
+#
+#   * create GCE host
+#   * provision staging VMs within GCE host
+#   * clean up GCE host
+#
+# We're using a wrapper script for this functionality to ensure
+# that errors propagate up to the CI runner, so that any non-zero
+# exit aborts the run and the CI job is marked as failing.
+set -e
+set -u
+set -o pipefail
+
+
+# Assume we're running against Trusty; Xenial also supported.
+target_platform="${1:trusty}"
+
+# Skip full CI run on docs branches, since code shouldn't change.
+if [[ "${CIRCLE_BRANCH:-}" != docs-* ]]; then
+    ./devops/gce-nested/gce-start.sh
+    ./devops/gce-nested/gce-runner.sh "$target_platform"
+    ./devops/gce-nested/gce-stop.sh
+else
+    echo "Branch begins with 'docs-' prefix, skipping staging tests..."
+fi

--- a/devops/gce-nested/ci-go.sh
+++ b/devops/gce-nested/ci-go.sh
@@ -14,7 +14,7 @@ set -o pipefail
 
 
 # Assume we're running against Trusty; Xenial also supported.
-target_platform="${1:trusty}"
+target_platform="${1:-trusty}"
 
 # Skip full CI run on docs branches, since code shouldn't change.
 if [[ "${CIRCLE_BRANCH:-}" != docs-* ]]; then

--- a/devops/gce-nested/gce-runner.sh
+++ b/devops/gce-nested/gce-runner.sh
@@ -16,6 +16,8 @@ REMOTE_IP="$(gcloud_call compute instances describe \
 SSH_TARGET="${SSH_USER_NAME}@${REMOTE_IP}"
 SSH_OPTS=(-i "$SSH_PRIVKEY" -o "StrictHostKeyChecking=no" -o "UserKnownHostsFile=/dev/null")
 
+# Assume we're testing Trusty. Xenial is also supported
+target_platform="${1:-trusty}"
 
 # Wrapper utility to run commands on remote GCE instance
 function ssh_gce {
@@ -53,7 +55,7 @@ function copy_securedrop_repo() {
 
 # Main logic
 copy_securedrop_repo
-if [[ "${1:-trusty}" = "xenial" ]]; then
+if [[ "$target_platform" = "xenial" ]]; then
     ssh_gce "make build-debs-xenial"
 else
     ssh_gce "make build-debs-notest"
@@ -65,7 +67,7 @@ trap fetch_junit_test_results EXIT
 
 # Run staging environment. If xenial is passed as an argument, run the
 # staging environment for xenial.
-if [[ "${1:-trusty}" = "xenial" ]]; then
+if [[ "$target_platform" = "xenial" ]]; then
     ssh_gce "make staging-xenial"
 else
     ssh_gce "make staging"

--- a/docs/development/testing_continuous_integration.rst
+++ b/docs/development/testing_continuous_integration.rst
@@ -89,9 +89,8 @@ Run ``make help`` to see the full list of CI commands in the Makefile:
     Makefile for developing and testing SecureDrop.
     Subcommands:
         ci-go                      Creates, provisions, tests, and destroys GCE host for testing staging environment.
+        ci-go-xenial               Creates, provisions, tests, and destroys GCE host for testing staging environment under xenial.
         ci-lint                    Runs linting in linting container.
-        ci-run                     Provisions GCE host for testing staging environment.
-        ci-spinup                  Creates GCE host for testing staging environment.
         ci-teardown                Destroys GCE host for testing staging environment.
 
 To run the tests locally:
@@ -100,6 +99,6 @@ To run the tests locally:
 
     make ci-go
 
-You can use ``make ci-run`` to provision the remote hosts while making changes,
-including rebuilding the Debian packages used in the Staging environment.
-See :doc:`virtual_environments` for more information.
+You can use ``./devops/gce-nested/ci-runner.sh`` to provision the remote hosts
+while making changes, including rebuilding the Debian packages used in the
+Staging environment. See :doc:`virtual_environments` for more information.

